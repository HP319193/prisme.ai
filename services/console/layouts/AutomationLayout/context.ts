import { createContext, useContext } from "react";
import { Workspace } from "../../api/types";
import { ValidationError } from "../../utils/yaml";

export interface AutomationLayoutContext {
  automation: Prismeai.Automation;
  setAutomation: (automation: Prismeai.Automation) => void;
  reset: () => void;
  save: () => void;
  invalid: false | ValidationError[];
}

export const automationLayoutContext = createContext<AutomationLayoutContext>({
  automation: {
    id: "",
    name: "",
    triggers: {},
    workflows: {},
  },
  setAutomation() {},
  reset() {},
  save() {},
  invalid: false,
});

export const useAutomation = () => useContext(automationLayoutContext);

export default automationLayoutContext;
