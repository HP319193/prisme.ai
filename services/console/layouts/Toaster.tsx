import {
  createContext,
  FC,
  useContext,
  useEffect,
  useRef,
  useState,
} from "react";
import { Toast } from "primereact/toast";

export const context = createContext<Toast>({} as Toast);

export const useToaster = () => useContext(context);

export const Toaster: FC = ({ children }) => {
  const ref = useRef(null);
  const [toaster, setToaster] = useState<any>();
  useEffect(() => {
    // Force toaster to re-render with ref
    setToaster(ref.current);
  }, [ref]);
  return (
    <context.Provider value={toaster!}>
      {children}
      <Toast ref={ref} position="bottom-right" />
    </context.Provider>
  );
};

export default Toaster;
