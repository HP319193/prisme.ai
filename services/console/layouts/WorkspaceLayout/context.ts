import { createContext, FC, useContext } from 'react';
import { Event, EventsFilters, Workspace } from '@prisme.ai/sdk';
import { ValidationError } from '../../utils/yaml';

export type Pagination = {
  page: PrismeaiAPI.EventsLongpolling.Parameters.Page;
  limit: PrismeaiAPI.EventsLongpolling.Parameters.Limit;
};
export type EventsByDay = Map<number, Set<Event<Date>>>;
export interface WorkspaceContext {
  displaySource: (status: boolean) => void;
  sourceDisplayed: boolean;
  invalid: false | ValidationError[];
  setInvalid: (invalid: WorkspaceContext['invalid']) => void;
  newSource?: Workspace;
  setNewSource: (fn: WorkspaceContext['newSource']) => void;
  fullSidebar: boolean;
  setFullSidebar: (s: boolean) => void;

  // To move into a WorkspaceProvider
  workspace: Workspace;
  loading: boolean;
  filters: EventsFilters;
  updateFilters: (newFilters: EventsFilters) => void;
  save: (workspace: Workspace) => void;
  saveSource: () => void;
  saving: boolean;
  events: EventsByDay | 'loading';
  nextEvents: () => void;
  readEvents: Set<string>;
  readEvent: (eventId: string) => void;
  share?: {
    label: string;
    component: FC;
  };
  setShare: (share: WorkspaceContext['share']) => void;
  getAppConfig: (appInstance: string) => any;
  saveAppConfig: (appInstance: string, config: any) => void;
  createAutomation: (
    automation: Prismeai.Automation
  ) => Promise<(Prismeai.Automation & { slug: string }) | null>;
  updateAutomation: (
    slug: string,
    automation: Prismeai.Automation
  ) => Promise<(Prismeai.Automation & { slug: string }) | null>;
  deleteAutomation: (slug: string) => Promise<Prismeai.Automation | null>;
}

export const workspaceContext = createContext<WorkspaceContext>({
  displaySource() {},
  sourceDisplayed: false,
  invalid: false,
  setInvalid() {},
  setNewSource() {},
  fullSidebar: false,
  setFullSidebar() {},
  //
  workspace: {} as Workspace,
  loading: false,
  filters: {} as EventsFilters,
  updateFilters: () => {},
  save() {},
  saveSource() {},
  saving: false,
  events: 'loading',
  nextEvents() {},
  readEvents: new Set(),
  readEvent() {},
  setShare() {},
  getAppConfig() {},
  saveAppConfig() {},
  createAutomation: async () => ({} as Prismeai.Automation & { slug: string }),
  updateAutomation: async () => ({} as Prismeai.Automation & { slug: string }),
  deleteAutomation: async () => ({} as Prismeai.Automation),
});

export const useWorkspace = () => useContext(workspaceContext);

export default workspaceContext;
