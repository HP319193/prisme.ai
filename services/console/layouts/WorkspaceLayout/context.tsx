import { createContext } from 'react';
import { Workspace } from '@prisme.ai/sdk';
import { ValidationError } from '../../utils/yaml';
import { useContext } from '../../utils/useContext';

export enum DisplayedSourceType {
  Config = 'config',
  Roles = 'roles',
  None = 'none',
}
export interface WorkspaceLayoutContext {
  displaySource: (type: DisplayedSourceType) => void;
  sourceDisplayed: DisplayedSourceType;
  invalid: false | ValidationError[];
  setInvalid: (invalid: WorkspaceLayoutContext['invalid']) => void;
  saving: boolean;
  setSaving: (s: WorkspaceLayoutContext['saving']) => void;
  onSave: (workspace: Workspace) => void;
  onSaveSource: () => void;
  newSource?: Workspace;
  setNewSource: (fn: WorkspaceLayoutContext['newSource']) => void;
  fullSidebar: boolean;
  setFullSidebar: (s: boolean) => void;
  createAutomation: (
    values: Pick<Prismeai.Automation, 'slug' | 'name'>
  ) => Promise<boolean>;
  createPage: (
    options?: { slug?: string; public?: true } & Partial<Prismeai.Page>
  ) => Promise<boolean>;
  installApp: () => void;
  createBlock: (values: {
    slug: string;
    name: Prismeai.LocalizedText;
  }) => Promise<boolean>;
  advancedMode: boolean;
  setAdvancedMode: (advancedMode: boolean) => void;
}

export const workspaceLayoutContext = createContext<
  WorkspaceLayoutContext | undefined
>(undefined);

export const useWorkspaceLayout = () =>
  useContext<WorkspaceLayoutContext>(workspaceLayoutContext);

export default workspaceLayoutContext;
