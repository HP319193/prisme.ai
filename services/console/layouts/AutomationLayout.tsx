import { FC, ReactElement, useCallback, useMemo } from "react";
import { TabMenu } from "primereact/tabmenu";
import { useWorkspace } from "./WorkspaceLayout";
import { useRouter } from "next/router";
import EditableTitle from "../components/EditableTitle";
import { useWorkspaces } from "../components/WorkspacesProvider";

export const AutomationLayout: FC = ({ children }) => {
  const { workspace } = useWorkspace();
  const { update } = useWorkspaces();
  const {
    query: { id, name },
    route,
    push,
    replace,
  } = useRouter();
  const currentTab = route.split(/\//).pop();
  const updateTitle = useCallback(
    async (newTitle: string) => {
      if (newTitle === name) return;
      if (Object.keys(workspace.automations).includes(newTitle)) return;
      const newAutomations = { ...workspace.automations };
      newAutomations[newTitle] = { ...newAutomations[`${name}`] };
      delete newAutomations[`${name}`];
      await update({
        ...workspace,
        automations: newAutomations,
      });
      replace(`/workspaces/${id}/automations/${newTitle}/${currentTab}`);
    },
    [currentTab, id, name, replace, update, workspace]
  );
  const title = useMemo(
    () => [
      {
        label: "back",
        className: "",
        command: () => push(`/workspaces/${workspace.id}`),
      },
      {
        className: "flex justify-content-center align-items-center ml-2",
        template: () => (
          <EditableTitle title={`${name}`} onChange={updateTitle} />
        ),
      },
    ],
    [name, push, updateTitle, workspace.id]
  );
  const tabs = useMemo(
    () => [
      {
        label: "Design",
        command: () =>
          push(`/workspaces/${workspace.id}/automations/${name}/design`),
      },
      {
        label: "Manifest",
        command: () =>
          push(`/workspaces/${workspace.id}/automations/${name}/manifest`),
      },
    ],
    [name, push, workspace.id]
  );

  return (
    <div className="flex flex-column ">
      <div className="flex flex-1 flex-row">
        <TabMenu model={title} activeIndex={-1} className="flex-1" />
        <TabMenu model={tabs} activeIndex={currentTab === "design" ? 0 : 1} />
      </div>
      <div className="flex">{children}</div>
    </div>
  );
};

export const getLayout = (page: ReactElement) => (
  <AutomationLayout>{page}</AutomationLayout>
);

export default AutomationLayout;
