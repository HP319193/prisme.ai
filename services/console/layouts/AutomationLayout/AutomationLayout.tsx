import { FC, useCallback, useEffect, useMemo, useState } from "react";
import { TabMenu } from "primereact/tabmenu";
import { useWorkspace } from "../WorkspaceLayout";
import { useRouter } from "next/router";
import EditableTitle from "../../components/EditableTitle";
import { useWorkspaces } from "../../components/WorkspacesProvider";
import { useTranslation } from "next-i18next";
import { Button } from "primereact/button";
import context, { AutomationLayoutContext } from "./context";
import { validateAutomation } from "@prisme.ai/validation";
import { ValidationError } from "../../utils/yaml";
import Error404 from "../../views/Errors/404";
import { useToaster } from "../Toaster";

export const AutomationLayout: FC = ({ children }) => {
  const { t } = useTranslation("workspaces");
  const {
    query: { automationId },
    route,
    push,
  } = useRouter();
  const { workspace } = useWorkspace();
  const { updateAutomation } = useWorkspaces();
  const [automation, setCurrentAutomation] = useState<Prismeai.Automation>();
  const toaster = useToaster();
  const [dirty, setDirty] = useState(false);

  const [invalid, setInvalid] = useState<false | ValidationError[]>(false);

  const currentTab = route.split(/\//).pop();

  const reset: AutomationLayoutContext["reset"] = useCallback(() => {
    setCurrentAutomation(
      workspace.automations.find((automation) => automation.id === automationId)
    );
    setDirty(false);
  }, [automationId, workspace]);
  useEffect(() => {
    reset();
  }, [reset]);

  const updateTitle = useCallback(
    async (newTitle: string) => {
      if (!automation || automation.name === newTitle) return;
      setDirty(true);
      updateAutomation(workspace, {
        ...automation,
        name: newTitle,
      });
    },
    [automation, updateAutomation, workspace]
  );

  const setAutomation: AutomationLayoutContext["setAutomation"] = useCallback(
    (newAutomation) => {
      validateAutomation(newAutomation);
      validateAutomation.errors;
      setInvalid((validateAutomation.errors as ValidationError[]) || false);
      setCurrentAutomation({
        ...automation,
        ...newAutomation,
      });
      setDirty(true);
    },
    [automation]
  );

  const save: AutomationLayoutContext["save"] = useCallback(async () => {
    if (invalid || !automation) return;
    try {
      await updateAutomation(workspace, {
        ...automation,
      });
      toaster.show({
        severity: "success",
        summary: t("automations.save.toast"),
      });
      setDirty(false);
    } catch (e) {
      toaster.show({
        severity: "error",
        summary: t("automations.save.error"),
      });
    }
  }, [invalid, automation, toaster, t, updateAutomation, workspace]);

  const title = useMemo(
    () => [
      {
        label: t("automations.back"),
        className: "",
        command: () => push(`/workspaces/${workspace.id}`),
      },
      {
        className: "flex justify-content-center align-items-center ml-2",
        template: () => (
          <EditableTitle
            title={`${automation && automation.name}`}
            onChange={updateTitle}
          />
        ),
      },
    ],
    [automation, push, t, updateTitle, workspace.id]
  );

  const tabs = useMemo(
    () => [
      {
        label: t("automations.design.title"),
        command: () =>
          push(
            `/workspaces/${workspace.id}/automations/${automationId}/design`
          ),
      },
      {
        label: t("automations.manifest.title"),
        command: () =>
          push(
            `/workspaces/${workspace.id}/automations/${automationId}/manifest`
          ),
      },
      {
        className: "flex justify-content-center align-items-center ml-2 mr-2",
        template: () => (
          <div>
            <Button onClick={save} disabled={!dirty || !!invalid}>
              {t("automations.save.label")}
            </Button>
          </div>
        ),
      },
    ],
    [automationId, dirty, invalid, push, save, t, workspace.id]
  );

  if (!automation) {
    return <Error404 link="/automations" reason={t("404")} />;
  }

  return (
    <context.Provider
      value={{ automation, setAutomation, reset, save, invalid }}
    >
      <div className="flex flex-column flex-1">
        <div className="flex flex-row">
          <TabMenu model={title} activeIndex={-1} className="flex-1" />
          <TabMenu model={tabs} activeIndex={currentTab === "design" ? 0 : 1} />
        </div>
        <div className="flex flex-1 flex-column">{children}</div>
      </div>
    </context.Provider>
  );
};

export default AutomationLayout;
