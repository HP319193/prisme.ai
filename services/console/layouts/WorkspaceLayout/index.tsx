import { ReactElement } from 'react';
import WorkspaceLayout from './WorkspaceLayout';

export * from './context';
export * from './WorkspaceLayout';

export const getLayout = (page: ReactElement) => {
  return <WorkspaceLayout>{page}</WorkspaceLayout>;
};

export default getLayout;
