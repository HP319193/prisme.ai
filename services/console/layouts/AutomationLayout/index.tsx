import { ReactElement } from "react";
import AutomationLayout from "./AutomationLayout";

export * from "./context";
export * from "./AutomationLayout";

export const getLayout = (page: ReactElement) => (
  <AutomationLayout>{page}</AutomationLayout>
);

export default getLayout;
