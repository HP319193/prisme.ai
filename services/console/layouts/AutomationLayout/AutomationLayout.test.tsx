import AutomationLayout from "./AutomationLayout";
import getLayout from "./index";
import renderer, { act } from "react-test-renderer";
import { useRouter } from "next/router";
import { TabMenu } from "primereact/tabmenu";
import EditableTitle from "../../components/EditableTitle";
import { useWorkspaces } from "../../components/WorkspacesProvider";
import { useAutomation } from "./context";
import { useWorkspace } from "../WorkspaceLayout";
import { Button } from "primereact/button";
import { useToaster } from "../Toaster";

jest.mock("primereact/tabmenu", () => ({
  TabMenu: () => null,
}));
jest.mock("../WorkspaceLayout", () => {
  const workspace = {
    id: "42",
  };
  return {
    useWorkspace: () => ({
      workspace,
    }),
  };
});

jest.mock("next/router", () => {
  const mock = {
    query: { id: "42", automationId: "43" },
    route: "/workspaces/42/automations/43/manifest",
    push: jest.fn(),
    replace: jest.fn(),
  };
  return {
    useRouter: () => mock,
  };
});

jest.mock("../../components/WorkspacesProvider", () => {
  const mock = {
    updateAutomation: jest.fn(),
  };
  return {
    useWorkspaces: () => mock,
  };
});

jest.mock("../Toaster", () => {
  const mock = {
    show: jest.fn(),
  };
  return { useToaster: () => mock };
});

beforeEach(() => {
  useWorkspace().workspace.automations = [
    {
      id: "43",
      name: "foo",
      triggers: {},
      workflows: {},
    },
    {
      id: "44",
      name: "foofoo",
      triggers: {},
      workflows: {},
    },
  ];
  (useWorkspaces().updateAutomation as jest.Mock).mockClear();
});

it("should render", () => {
  const root = renderer.create(<AutomationLayout>Foo</AutomationLayout>);
  expect(root.toJSON()).toMatchSnapshot();
});

it("should render with manifest tab active", () => {
  useRouter().route = "/workspaces/42/automations/foo/manifest";
  const root = renderer.create(<AutomationLayout>Foo</AutomationLayout>);
  expect(root.toJSON()).toMatchSnapshot();
});

it("should navigate", async () => {
  const root = renderer.create(<AutomationLayout>Foo</AutomationLayout>);
  await act(async () => {
    await true;
  });
  act(() => {
    root.root.findAllByType(TabMenu)[0].props.model[0].command();
  });
  expect(useRouter().push).toHaveBeenCalledWith("/workspaces/42");
  (useRouter().push as any).mockClear();

  act(() => {
    root.root.findAllByType(TabMenu)[1].props.model[0].command();
  });
  expect(useRouter().push).toHaveBeenCalledWith(
    "/workspaces/42/automations/43/design"
  );
  (useRouter().push as any).mockClear();

  act(() => {
    root.root.findAllByType(TabMenu)[1].props.model[1].command();
  });
  expect(useRouter().push).toHaveBeenCalledWith(
    "/workspaces/42/automations/43/manifest"
  );
});

it("should edit title", async () => {
  const root = renderer.create(<AutomationLayout>Foo</AutomationLayout>);
  await act(async () => {
    await true;
  });
  const itemRoot = renderer.create(
    root.root.findAllByType(TabMenu)[0].props.model[1].template()
  );
  expect(itemRoot.toJSON()).toMatchSnapshot();

  await act(async () => {
    await itemRoot.root.findByType(EditableTitle).props.onChange("foo");
  });

  const { updateAutomation } = useWorkspaces();

  expect(updateAutomation).not.toHaveBeenCalled();

  await act(async () => {
    await itemRoot.root.findByType(EditableTitle).props.onChange("foofoo");
  });

  await act(async () => {
    await itemRoot.root.findByType(EditableTitle).props.onChange("bar");
  });
});

it("should get layout", async () => {
  const root = renderer.create(getLayout(<div />));
  await act(async () => {
    await true;
  });
  expect(root.toJSON()).toMatchSnapshot();
});

it("should set and reset automation", async () => {
  let context: any;
  const Test = () => {
    context = useAutomation();
    return null;
  };
  const root = renderer.create(
    <AutomationLayout>
      <Test />
    </AutomationLayout>
  );
  await act(async () => {
    await true;
  });
  const expectedAutomation = {
    triggers: {},
    workflows: {
      something: {
        do: [
          {
            emit: {
              events: ["event"],
            },
          },
        ],
      },
    },
  };
  act(() => {
    context.setAutomation(expectedAutomation);
  });

  expect(context.automation).toEqual({
    id: "43",
    name: "foo",
    ...expectedAutomation,
  });

  act(() => {
    context.reset();
  });

  expect(context.automation).toEqual({
    id: "43",
    name: "foo",
    triggers: {},
    workflows: {},
  });
});

it("should save automation", async () => {
  let context: any;
  const Test = () => {
    context = useAutomation();
    return null;
  };
  const root = renderer.create(
    <AutomationLayout>
      <Test />
    </AutomationLayout>
  );
  await act(async () => {
    await true;
  });
  await act(async () => {
    await context.save();
  });
  expect(useWorkspaces().updateAutomation).toHaveBeenCalledWith(
    useWorkspace().workspace,
    {
      id: "43",
      name: "foo",
      triggers: {},
      workflows: {},
    }
  );
});

it("should save", async () => {
  const root = renderer.create(<AutomationLayout>foo</AutomationLayout>);
  await act(async () => {
    await true;
  });
  const itemRoot = renderer.create(
    root.root.findAllByType(TabMenu)[1].props.model[2].template()
  );
  await act(async () => {
    await itemRoot.root.findByType(Button).props.onClick();
  });
  expect(useWorkspaces().updateAutomation).toHaveBeenCalled();
});

it("should fail to save", async () => {
  const root = renderer.create(<AutomationLayout>foo</AutomationLayout>);
  await act(async () => {
    await true;
  });
  (useWorkspaces().updateAutomation as jest.Mock).mockImplementationOnce(() => {
    throw new Error();
  });
  const itemRoot = renderer.create(
    root.root.findAllByType(TabMenu)[1].props.model[2].template()
  );
  await act(async () => {
    await itemRoot.root.findByType(Button).props.onClick();
  });
  expect(useToaster().show).toHaveBeenCalled();
});
