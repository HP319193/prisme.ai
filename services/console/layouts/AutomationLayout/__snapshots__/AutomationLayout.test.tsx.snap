// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should edit title 1`] = `
<div
  className="p-inplace p-component p-inplace-closable flex"
>
  <div
    aria-label={null}
    className="p-inplace-display"
    onClick={[Function]}
    onKeyDown={[Function]}
    tabIndex={0}
  >
    <div
      className="flex"
    >
      foo
    </div>
  </div>
</div>
`;

exports[`should get layout 1`] = `
<div
  className="flex flex-column flex-1"
>
  <div
    className="flex flex-row"
  />
  <div
    className="flex flex-1 flex-column"
  >
    <div />
  </div>
</div>
`;

exports[`should render 1`] = `
<a
  href="/automations"
  onClick={[Function]}
  onMouseEnter={[Function]}
>
  404
</a>
`;

exports[`should render with manifest tab active 1`] = `
<div
  className="flex flex-column flex-1"
>
  <div
    className="flex flex-row"
  />
  <div
    className="flex flex-1 flex-column"
  >
    Foo
  </div>
</div>
`;
